File: fragments/api/fastify/datastore/redis.ts
import { type FastifyInstance, type FastifyPluginOptions } from "fastify";
import { redisCreate, redisDelete, redisPing, redisRead, redisUpdate } from "../../../redis_db/redis_crud";

export const routes = (fastify: FastifyInstance, _: FastifyPluginOptions) => {
  fastify.get<{
    Reply: {
      200: { message: string };
      500: { error: string };
    };
  }>("/ping", async (request, reply) => {
    try {
      const result = await redisPing();
      return { message: result };
    } catch (error) {
      request.log.error(error, "Error pinging Redis");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });

  fastify.get<{
    Params: { namespace: string };
    Reply: {
      200: { favouriteCoin?: string };
      404: void;
      500: { error: string };
    };
  }>("/favourites/:namespace", async (request, reply) => {
    try {
      const { namespace } = request.params;
      const item = await redisRead(namespace);

      if (!item.favouriteCoin) {
        return reply.code(404).send();
      }
      return item;
    } catch (error) {
      request.log.error(error, "Error reading favourite coin from Redis");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });

  fastify.put<{
    Body: { favouriteCoin: string };
    Params: { namespace: string };
    Reply: {
      200: void;
      400: { error: string };
      500: { error: string };
    };
  }>("/favourites/:namespace", {
    schema: {
      body: {
        type: "object",
        required: ["favouriteCoin"],
        properties: {
          favouriteCoin: { type: "string" },
        },
      },
    },
  }, async (request, reply) => {
    try {
      const { namespace } = request.params;
      const { favouriteCoin } = request.body;
      const result = await redisCreate(namespace, favouriteCoin);
      if (result !== "OK") {
        throw new Error(result);
      }
      return reply.code(200).send();
    } catch (error) {
      request.log.error(error, "Error creating favourite coin in Redis");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });

  fastify.patch<{
    Body: { favouriteCoin: string };
    Params: { namespace: string };
    Reply: {
      200: void;
      500: { error: string };
    };
  }>("/favourites/:namespace", {
    schema: {
      body: {
        type: "object",
        required: ["favouriteCoin"],
        properties: {
          favouriteCoin: { type: "string" },
        },
      },
    },
  }, async (request, reply) => {
    try {
      const { namespace } = request.params;
      const { favouriteCoin } = request.body;
      const result = await redisUpdate(namespace, favouriteCoin);
      if (result !== "OK") {
        throw new Error(result);
      }
      return reply.code(200).send();
    } catch (error) {
      request.log.error(error, "Error updating favourite coin in Redis");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });

  fastify.delete<{
    Params: { namespace: string };
    Reply: {
      204: void;
      500: { error: string };
    };
  }>("/favourites/:namespace", async (request, reply) => {
    try {
      const { namespace } = request.params;
      const result = await redisDelete(namespace);
      if (result !== "OK") {
        throw new Error(result);
      }
      return reply.code(204).send();
    } catch (error) {
      request.log.error(error, "Error deleting favourite coin from Redis");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });
};

File: fragments/apis/fastapi/datastore/redis.py
from typing import TypedDict
from fastapi import APIRouter, HTTPException, Response
from pydantic import BaseModel

from fragments.redis_db.redis_crud import (
    redis_ping,
    redis_create,
    redis_read,
    redis_update,
    redis_delete,
)

redis_router = APIRouter()


class PingResponse(TypedDict):
    message: str


class FavouriteResponse(TypedDict):
    favourite_coin: str


class FavouriteBody(BaseModel):
    favourite_coin: str


@redis_router.get("/ping", status_code=200)
async def ping() -> PingResponse:
    result = redis_ping()
    return PingResponse(message=result)


@redis_router.get("/favourites/{namespace}", status_code=200)
async def get_favourite(namespace: str) -> FavouriteResponse:
    item = redis_read(namespace)
    if "favourite_coin" not in item:
        raise HTTPException(status_code=404)
    return FavouriteResponse(favourite_coin=item["favourite_coin"])


@redis_router.put("/favourites/{namespace}", status_code=200)
async def create_favourite(namespace: str, body: FavouriteBody):
    redis_create(namespace, body.favourite_coin)
    return None


@redis_router.patch("/favourites/{namespace}", status_code=200)
async def update_favourite(namespace: str, body: FavouriteBody):
    redis_update(namespace, body.favourite_coin)
    return None


@redis_router.delete("/favourites/{namespace}", status_code=204)
async def delete_favourite(namespace: str):
    redis_delete(namespace)
    return Response(status_code=204)
